system/lib/pthread/library_pthread.c

```diff
@@ -194,47 +195,54 @@ void emscripten_async_waitable_close(em_queued_call *call)
+			case EM_FUNC_SIG_VIIIIII: ((em_func_viiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i); break;
+			case EM_FUNC_SIG_VIIIIIII: ((em_func_viiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i); break;
+			case EM_FUNC_SIG_VIIIIIIII: ((em_func_viiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i); break;
+			case EM_FUNC_SIG_VIIIIIIIII: ((em_func_viiiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i, q->args[8].i); break;
+			case EM_FUNC_SIG_VIIIIIIIIII: ((em_func_viiiiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i, q->args[8].i, q->args[9].i); break;
+			case EM_FUNC_SIG_VIIIIIIIIIII: ((em_func_viiiiiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i, q->args[8].i, q->args[9].i, q->args[10].i); break;
+			case EM_FUNC_SIG_I: q->returnValue.i = ((em_func_i)q->functionPtr)(); break;
+			case EM_FUNC_SIG_II: q->returnValue.i = ((em_func_ii)q->functionPtr)(q->args[0].i); break;
+			case EM_FUNC_SIG_III: q->returnValue.i = ((em_func_iii)q->functionPtr)(q->args[0].i, q->args[1].i); break;
+			case EM_FUNC_SIG_IIII: q->returnValue.i = ((em_func_iiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i); break;
+			case EM_FUNC_SIG_IIIII: q->returnValue.i = ((em_func_iiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i); break;
+			case EM_FUNC_SIG_IIIIII: q->returnValue.i = ((em_func_iiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i); break;
+			case EM_FUNC_SIG_IIIIIII: q->returnValue.i = ((em_func_iiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i); break;
+			case EM_FUNC_SIG_IIIIIIII: q->returnValue.i = ((em_func_iiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i); break;
+			case EM_FUNC_SIG_IIIIIIIII: q->returnValue.i = ((em_func_iiiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i); break;
+			case EM_FUNC_SIG_IIIIIIIIII: q->returnValue.i = ((em_func_iiiiiiiiii)q->functionPtr)(q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i, q->args[8].i); break;
+			default: assert(0 && "Invalid Emscripten pthread _do_call opcode!");
+		}
+	} else {
+		// JS call
+		q->returnValue.d = EM_ASM_DOUBLE({
+			return receiveOnMainThread($0);
+		}, q->js);
```

Comment:
Instead of this, let's move `receiveOnMainThread()` out from `preamble.js` to `library_pthread.js`, and declare it e.g. as `double runJavaScriptFunctionByIndex(int index, double* args);` to it to avoid the `EM_ASM_DOUBLE()` construct?

Reply:
Or perhaps even better might be

```c
double runJavaScriptFunctionByIndex(int index, int numArgs, double* args);
```

```js
function runJavaScriptFunctionByIndex(index, numArgs, args) {
  var func = index > 0 ? proxiedFunctionTable[index] : ASM_CONSTS[-index - 1];
#if SMALL_PTHREAD_PROXIED_FUNCTION_DISPATCH
  var callArgs = HEAPF64.subarray((args >> 3) + 2, (args >> 3) + 2 + numArgs);
  return func.apply(null, callArgs);
#else
  var a = args>>3;
  switch(numArgs) {
    case 0: return func();
    case 1: return func(HEAPF64[a]);
    case 2: return func(HEAPF64[a], HEAPF64[a+1]);
    case 3: return func(HEAPF64[a], HEAPF64[a+1], HEAPF64[a+2]);
    ...
    case 9: return func(HEAPF64[a], HEAPF64[a+1], HEAPF64[a+2], HEAPF64[a+3], HEAPF64[a+4], HEAPF64[a+5], HEAPF64[a+6], HEAPF64[a+7], HEAPF64[a+8]);
#if ASSERTIONS
    default: abort();
#endif
  }
#endif
}
```

That would retain garbage-free operation and avoiding `.apply()` to be able to compare. And then instead of using a `void* js;` parameter in `em_queued_call` object with a buffer, we could define a special `functionEnum` for JS functions:

```
#define EM_PROXIED_JS_FUNCTION(numArgs) (EM_PROXIED_FUNC_SPECIAL(4) | EM_FUNC_SIG_WITH_N_PARAMETERS(numArgs))
```

and in `em_queued_call` struct `functionPtr` could denote the index of the function. Then instead of 

```
q->returnValue.d = EM_ASM_DOUBLE({
			return receiveOnMainThread($0);
		}, q->js);
```

and a `if(q->js) ... else ... ` block, we could have

```
if ((q->functionEnum & ~EM_FUNC_SIG_NUM_PARAMETERS_MASK) == EM_PROXIED_JS_FUNCTION(0)) {
  q->returnValue.d = runJavaScriptFunctionByIndex(q->functionPtr, EM_FUNC_SIG_NUM_FUNC_ARGUMENTS(q->functionEnum), (double*)q->args);
} else switch(q->functionEnum) {
   ...
}
```

(it looks like the `EM_FUNC_SIG_NUM_PARAMETERS_MASK` could even be ANDed off of `q->functionEnum` for the switch-case check and in all case blocks, then the above if() test could be subsumed inside the `switch-case` altogether)

Reply:
I agree a JS library function is nice in that we can call it directly, avoiding an `EM_ASM`. On the other hand, I think there is value in having `proxy*` and `receive*` alongside each other in the code, it's simple and clear to read and understand the ABI that way. How about if I make both JS library functions? Adding a commit with that.

Reply:
Having both as library functions does sound good.